### PR TITLE
Prepopulate mongodb

### LINKED ISSUE #12: Put UCSUSA data into mongo
Put https://www.ucsusa.org/resources/satellite-database data into MongoDB.
You need #6 to be done first before starting with this task.

### PR DESCRIPTION
**Pull request by jpitucha**
resolves #12

### CODE REVIEW

File: src/utils/satelliteUtils.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import mongoose from 'mongoose'
+
+export default class SateliteUtilities {
+
+    static hasUCSData(): Promise<number> {
+        const satelite = mongoose.connection.db.collection('satelites')
+        return new Promise((resolve, reject) => {
+            satelite.estimatedDocumentCount((err, count) => {
+                if (err) return reject()
+                console.log('has ucs data', count)
```
Comment (Jordan141) [inline src/utils/satelliteUtils.ts]: Remove console .log

File: src/utils/utils.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    static prePopulateDatabase(): Promise<void> {
+
+        return new Promise((resolve, reject): void => {
+
+            const headings = this.getUsableHeadingsFromFile()
+            const satelites = this.getSatelitesFromFile()
+
+            satelites.forEach((satelite: string[]) => {
+                const finalObj: Record<string, string> = {}
```
Comment (Jordan141) [inline src/utils/utils.ts]: Better name for finalObj

File: src/utils/utils.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                headings.forEach((heading, index) => { finalObj[heading] = satelite[index] ?? '' })
+
+                DatabaseUtilities.createSatelite(finalObj)
+                .then()
```
Comment (Jordan141) [inline src/utils/utils.ts]: Delete this .then

**Review by Jordan141 — CHANGES_REQUESTED** [review]

File: src/db/schema/satelite.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import mongoose from 'mongoose'
+import Utilities from '../../utils/utils'
+
+const headings = Utilities.getUsableHeadingsFromFile()
+const finalObj = Object.fromEntries(headings.map((h) => [h, String]))
```
Comment (Jordan141) [inline src/db/schema/satelite.ts]: Better name

File: src/db/dbConnectionProvider.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import mongoose from 'mongoose'
+
+const connectToDatabase = (): Promise<void> => {
+    return new Promise((resolve, reject) => {
+        if (!process.env.DB_URL) return reject()
+        mongoose.connect(process.env.DB_URL, { useNewUrlParser: true, useUnifiedTopology: true, useCreateIndex: true })
```
Comment (michaelKurowski) [inline src/db/dbConnectionProvider.ts]: Why do you encapsulate it inside a promise if it already returns then and catch?

File: src/db/schema/satelite.ts (hunk 4 above)
Comment (michaelKurowski) [reply src/db/schema/satelite.ts]: What does this line do

File: src/index.ts [hunk 6]
```diff
@@ -1,18 +1,35 @@
...
 
-    app.use('/', routes)
-
-    app.listen(process.env.PORT, () => {
-        console.log(`Houston running at http://localhost:${process.env.PORT}`)
+connectToDatabase()
+    .then(() => SatelliteUtils.hasUCSData())
+    .then((count) => {
+        if (count < Utilities.countSatelitesFromFile()) {
+            DatabaseUtilities.deleteAllSatelites()
```
Comment (michaelKurowski) [inline src/index.ts]: You nest promises

File: src/utils/dbUtils.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import mongoose from 'mongoose'
+
+export default class DatabaseUtilities {
+
+    static deleteAllSatelites(): Promise<void> {
+        const satelite = mongoose.connection.db.collection('satelites')
+        return new Promise((resolve, reject) => {
+            satelite.deleteMany({ }, ((err) => {
```
Comment (michaelKurowski) [inline src/utils/dbUtils.ts]: Isn't there an `.exec()` variant

File: src/utils/dbUtils.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+                if (err) return reject()
+            }))
+            return resolve()
+        })
+    }
+
+    static createSatelite(newSatelite: Record<string, string>): Promise<void> {
+        const satelite = mongoose.connection.db.collection('satelites')
+        return new Promise((resolve, reject) => {
+            satelite.insertOne(newSatelite)
```
Comment (michaelKurowski) [inline src/utils/dbUtils.ts]: You nest promises

File: src/utils/satelliteUtils.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import mongoose from 'mongoose'
+
+export default class SateliteUtilities {
+
+    static hasUCSData(): Promise<number> {
+        const satelite = mongoose.connection.db.collection('satelites')
+        return new Promise((resolve, reject) => {
+            satelite.estimatedDocumentCount((err, count) => {
```
Comment (michaelKurowski) [inline src/utils/satelliteUtils.ts]: Is there no `.exec()` variant

File: src/utils/utils.ts [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        if (!process.env.DB_URL) return false
+        return true 
+    }
+
+    static getUsableHeadingsFromFile(): string[] {
+        const data = fs.readFileSync('./ucs-satelite-db.txt', 'utf8')
+        if (!data) return []
+        const headings = data.split('\n')[0]?.split('\t')
+        const lastUsablePropIndex = headings?.findIndex(element => element == 'Comments') ?? 0
+        const finalHeadings = headings?.slice(0, lastUsablePropIndex)
```
Comment (michaelKurowski) [inline src/utils/utils.ts]: Why headings can not have splice

File: src/utils/utils.ts [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        if (!process.env.DB_PASSWORD) return false
+        if (!process.env.DB_URL) return false
+        return true 
+    }
+
+    static getUsableHeadingsFromFile(): string[] {
+        const data = fs.readFileSync('./ucs-satelite-db.txt', 'utf8')
+        if (!data) return []
+        const headings = data.split('\n')[0]?.split('\t')
+        const lastUsablePropIndex = headings?.findIndex(element => element == 'Comments') ?? 0
```
Comment (michaelKurowski) [inline src/utils/utils.ts]: Why headings might not have findIndex

File: src/utils/utils.ts [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+        return new Promise((resolve, reject): void => {
+
+            const headings = this.getUsableHeadingsFromFile()
+            const satelites = this.getSatelitesFromFile()
+
+            satelites.forEach((satelite: string[]) => {
+                const finalObj: Record<string, string> = {}
+                headings.forEach((heading, index) => { finalObj[heading] = satelite[index] ?? '' })
+
+                DatabaseUtilities.createSatelite(finalObj)
```
Comment (michaelKurowski) [inline src/utils/utils.ts]: You don't return this promise, is it on purpose

File: src/utils/utils.ts [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+                .then()
+                .catch((err) => { if (err) return reject() })
```
Comment (michaelKurowski) [inline src/utils/utils.ts]: This line does literally nothing

**Review by michaelKurowski — CHANGES_REQUESTED** [review]

File: src/db/schema/satelite.ts (hunk 4 above)
Comment (jpitucha) [reply src/db/schema/satelite.ts]: Its schema creation from 1 line of txt file, eg. { prop1: String, prop2: String etc }

File: src/utils/dbUtils.ts [hunk 14]
```diff
@@ -0,0 +1,14 @@
+import { Satellite } from '../db/schema/satellite'
+export default class DatabaseUtilities {
+
+    static async deleteAllSatelites(): Promise<void> {
+        await Satellite.deleteMany({ })
+    }
+
+    static async createSatelite(newSatelite: Record<string, string>): Promise<void> {
+        Satellite.create(newSatelite)
```
Comment (michaelKurowski) [inline src/utils/dbUtils.ts]: Why is this functon async if it doesn't return asynchronous code? Neither does it wait for async code inside

File: src/utils/dbUtils.ts [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+            .then(() => { return })
```
Comment (michaelKurowski) [inline src/utils/dbUtils.ts]: What does it do?

File: src/utils/dbUtils.ts [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+            .catch((err) => { console.log(err); throw 'error occured while putting new satellite' })
```
Comment (michaelKurowski) [inline src/utils/dbUtils.ts]: Don't do it an one-liner

File: src/db/dbConnectionProvider.ts [hunk 17]
```diff
@@ -0,0 +1,11 @@
+import mongoose from 'mongoose'
+export default class dbConnectionProvider {
+
+    static async connectToDatabase(): Promise<void> {
+        if (!process.env.DB_URL) return Promise.reject()
+        const conn = await mongoose.connect(process.env.DB_URL, { useNewUrlParser: true, useUnifiedTopology: true, useCreateIndex: true })
```
Comment (michaelKurowski) [inline src/db/dbConnectionProvider.ts]: Can't you just return this?

File: src/utils/utils.ts [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+    static prePopulateDatabase(): Promise<void> {
+
+        return new Promise((resolve, reject): void => {
+
+            const headings = this.getUsableHeadingsFromFile()
+            const satelites = this.getSatelitesFromFile()
+
+            satelites.forEach((satelite: string[]) => {
+                const sateliteDetails: Record<string, string> = {}
+                headings.forEach((heading, index) => { sateliteDetails[heading] = satelite[index] ?? '' })
```
Comment (michaelKurowski) [inline src/utils/utils.ts]:
```
const sateliteDetails: SatelliteInterface = headings.reduce((accumulator, heading, index) => {
         accumulator[heading] = satelite[index] ?? ''
         return accumulator
    }
, {})
```

Comment (michaelKurowski) [reply src/utils/utils.ts]:
```
const sateliteDetails = <SatelliteInterface> headings.reduce((accumulator, heading, index) => {
         accumulator[heading] = satelite[index] ?? ''
         return accumulator
    }
, {})
```

File: src/utils/dbUtils.ts (hunk 16 above)
Comment (michaelKurowski) [reply src/utils/dbUtils.ts]: Is it intentional to hide the error?

File: src/utils/satelliteUtils.ts [hunk 19]
```diff
@@ -0,0 +1,12 @@
+import { Satellite } from './../db/schema/satellite'
+export default class SatelliteUtilities {
+
+    static async getSateliteCount(): Promise<number> {
```
Comment (michaelKurowski) [inline src/utils/satelliteUtils.ts]: Get rid of async and await

File: src/utils/utils.ts [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+        if (!data) return 0
+        return data.split('\n').length - 2
+    }
+
+    static prePopulateDatabase(): Promise<void[]> {
+        const headings = this.satelliteHeaders
+        const satellites = this.getSatelitesFromFile()
+    
+        const satellitesCreation = satellites.map((satellite) => {
+            const satelliteDetails: Record<string, string> = {}
```
Comment (michaelKurowski) [inline src/utils/utils.ts]: new Satellite() basing on Satellite class that implement SatelliteInterface

File: src/utils/utils.ts [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+            headings.forEach((heading, index) => { satelliteDetails[heading] = satellite[index] ?? '' })
```
Comment (michaelKurowski) [inline src/utils/utils.ts]: you can throw this into the Satellite class constructor

Comment (michaelKurowski) [reply src/utils/utils.ts]: and then basically new Satellite(satellite)

Comment (michaelKurowski) [reply src/utils/utils.ts]: Also satellite needs different name, same for satelliteDetails

File: src/utils/satelliteUtils.ts [hunk 22]
```diff
@@ -0,0 +1,12 @@
...
+        return await Satellite.countDocuments()
+    }
+
+    static async removeCollectionIfExists(): Promise<void> {
+        return await Satellite.deleteMany({ }).exec()
```
Comment (michaelKurowski) [inline src/utils/satelliteUtils.ts]: async and await unecessary

File: src/utils/satelliteUtils.ts [hunk 23]
```diff
@@ -0,0 +1,20 @@
+import { Satellite } from './../db/schema/satellite'
+export default class SatelliteUtilities {
+
+    static createSatelite(newSatelite: Record<string, string>): void {
+        Satellite.create(newSatelite)
```
Comment (michaelKurowski) [inline src/utils/satelliteUtils.ts]: Do you not want to return this?

File: src/utils/utils.ts [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+        const usableDataMaxIndex = this.satelliteHeaders.length
+        const splittedSatelitesData: Array<string[]> = []
+        satelites.forEach((value) => { splittedSatelitesData.push(value.split('\t').slice(0, usableDataMaxIndex)) })
+        return splittedSatelitesData.slice(0, splittedSatelitesData.length -1)
+    }
+
+    static countSatelitesFromFile(): number {
+        const data = fs.readFileSync('./ucs-satelite-db.txt', 'utf8')
+        if (!data) return 0
+        return data.split('\n').length - 2
```
Comment (michaelKurowski) [inline src/utils/utils.ts:60]: Why -2?

File: src/utils/utils.ts [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+        type ConstructedSatellite = {
+            [P in SatelliteField] : string | number
+        }
+        const satellites = this.getSatelitesFromFile()
+    
+        const satellitesCreation = satellites.map((satellite) => {
+            const satelliteDetails = headings.reduce((accumulator, heading, index) => {
+                accumulator[heading] = satellite[index] ?? ''
+                return accumulator
+            }, {} as ConstructedSatellite)
```
Comment (michaelKurowski) [inline src/utils/utils.ts:75]: Isn't ConstructedSatellite same as SatelliteInterface?

File: src/db/schema/satelite.ts (hunk 4 above)
Comment (michaelKurowski) [reply src/db/schema/satelite.ts]: Call it with a name related to schema

File: src/utils/satelliteUtils.ts [hunk 26]
```diff
@@ -0,0 +1,21 @@
+import { Satellite } from './../db/schema/satellite'
+export default class SatelliteUtilities {
+
+    static async createSatelite(newSatelite: Record<string, string>): Promise<void> {
```
Comment (michaelKurowski) [inline src/utils/satelliteUtils.ts:4]: `async` and `await` unecessary. 
Just make it a normat function and do `return Satellite.create(newSatelite)`

**Review by michaelKurowski — APPROVED** [review]

File: src/utils/utils.ts [hunk 27]
```diff
@@ -0,0 +1,82 @@
...
+    static prePopulateDatabase(): Promise<void[]> {
+        const headings = this.satelliteHeaders
+        type SatelliteField = typeof headings[number]
+        type ConstructedSatellite = {
+            [P in SatelliteField] : string | number
+        }
+        const satellitesFromFile = this.getSatelitesFromFile()
+    
+        const satellitesCreation = satellitesFromFile.map((satellite) => {
+            const finalObj = headings.reduce((accumulator, heading, index) => {
```
Comment (Jordan141) [inline src/utils/utils.ts]: Give this a meaningful name

**Review by Jordan141 — CHANGES_REQUESTED** [review]

**Review by Jordan141 — APPROVED** [review]

**Pull request merged**